### PR TITLE
Optimize handling of purely-LTR paragraphs

### PR DESCRIPTION
**Pull request by jfkthame**
This aims to boost performance for paragraphs that do not contain any strongly-RTL characters.

Typical results of `cargo bench -F bench_it --bench udhr` with current master:

    test bench_1_bidi_info_new_for_ltr_texts  ... bench:     891,763 ns/iter (+/- 43,823)
    test bench_2_bidi_info_new_for_bidi_texts ... bench:   1,002,740 ns/iter (+/- 173,182)
    test bench_3_reorder_line_for_ltr_texts   ... bench:     155,455 ns/iter (+/- 17,878)
    test bench_4_reorder_line_for_bidi_texts  ... bench:     330,877 ns/iter (+/- 39,449)

And the same test after applying this patch:

    test bench_1_bidi_info_new_for_ltr_texts  ... bench:     875,862 ns/iter (+/- 99,252)
    test bench_2_bidi_info_new_for_bidi_texts ... bench:   1,005,841 ns/iter (+/- 179,152)
    test bench_3_reorder_line_for_ltr_texts   ... bench:         894 ns/iter (+/- 57)
    test bench_4_reorder_line_for_bidi_texts  ... bench:     243,300 ns/iter (+/- 16,616)

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -120,13 +120,19 @@ pub struct ParagraphInfo {
     ///
     /// <http://www.unicode.org/reports/tr9/#BD4>
     pub level: Level,
+
+    /// Whether any strong-RTL characters are present.
+    has_rtl: bool,
```
Comment (Manishearth) [inline src/lib.rs:125]: this is a breaking change: Paragraph could previously be constructed manually (and often is, due to the lack of convenient APIs here)

**Review by Manishearth — COMMENTED** [review]
I'm not sure how to make this work given that ParagraphInfo can be constructed from parts and this removes that ability

File: src/lib.rs (hunk 1 above)
Comment (jfkthame) [reply src/lib.rs:125]: Ah, that's a pity; does that need to be the case, or is it worth reconsidering?

(Note that `./cargo test` didn't complain for me, so I guess that isn't explicitly tested at present.)

Comment (Manishearth) [reply src/lib.rs:125]: I'd strongly like to avoid needing to do a semver breaking bump on a major dependency in the rust ecosystem.

**Comment by jfkthame** [comment]
I've updated this to take a different approach that (AFAICS) should not break any public API. @Manishearth PTAL and see if this is a workable way forward.

Current results on my local machine for `cargo bench -F bench_it --bench udhr` with master:
```
test bench_1_bidi_info_new_for_ltr_texts  ... bench:     672,764 ns/iter (+/- 94,785)
test bench_2_bidi_info_new_for_bidi_texts ... bench:     778,742 ns/iter (+/- 172,804)
test bench_3_reorder_line_for_ltr_texts   ... bench:     110,351 ns/iter (+/- 6,317)
test bench_4_reorder_line_for_bidi_texts  ... bench:     255,082 ns/iter (+/- 25,808)
```

Same test with this patch:
```
test bench_1_bidi_info_new_for_ltr_texts  ... bench:     215,393 ns/iter (+/- 23,973)
test bench_2_bidi_info_new_for_bidi_texts ... bench:     725,057 ns/iter (+/- 55,314)
test bench_3_reorder_line_for_ltr_texts   ... bench:       8,326 ns/iter (+/- 1,408)
test bench_4_reorder_line_for_bidi_texts  ... bench:     193,079 ns/iter (+/- 14,852)
```

So we get a pretty big improvement on the `ltr_texts` tests, and also a slight improvement on the `bidi_texts` ones (presumably the "bidi" texts do have a smattering of LTR-only paragraphs in them).

File: src/lib.rs [hunk 2]
```diff
@@ -175,17 +175,47 @@ impl<'text> InitialInfo<'text> {
...
+        let ext_info = InitialInfoExt::new_with_data_source(data_source, text, default_para_level);
+        ext_info.base
+    }
+}
+
+/// Extended version of InitialInfo (not public API).
+#[derive(PartialEq, Debug)]
+struct InitialInfoExt<'text> {
+    base: InitialInfo<'text>,
+    pure_ltr: Vec<bool>,
```
Comment (Manishearth) [inline src/lib.rs:187]: nit: please document this field

File: src/lib.rs [hunk 3]
```diff
@@ -175,17 +175,47 @@ impl<'text> InitialInfo<'text> {
...
         let mut original_classes = Vec::with_capacity(text.len());
 
         // The stack contains the starting byte index for each nested isolate we're inside.
         let mut isolate_stack = Vec::new();
         let mut paragraphs = Vec::new();
+        let mut pure_ltr = Vec::new();
 
         let mut para_start = 0;
         let mut para_level = default_para_level;
+        let mut is_pure_ltr = true;
```
Comment (Manishearth) [inline src/lib.rs:215]: nit: comment that this is per-para state

**Review by Manishearth — APPROVED** [review]

File: src/lib.rs (hunk 2 above)
Comment (Manishearth) [reply src/lib.rs:187]: hm, doesn't seem to have changed?

Comment (jfkthame) [reply src/lib.rs:187]: Errr.... I must have goofed, lemme try again....!

**Review by Manishearth — APPROVED** [review]

**Pull request merged**